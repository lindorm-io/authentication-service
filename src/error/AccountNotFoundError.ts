import { APIError, HttpStatus } from "@lindorm-io/core";

export class AccountNotFoundError extends APIError {
  constructor(accountId?: string) {
    super("Account not found", {
      debug: { accountId },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
