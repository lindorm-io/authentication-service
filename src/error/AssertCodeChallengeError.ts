import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertCodeChallengeError extends APIError {
  constructor(challenge: string, verifier: string) {
    super("Invalid Code Challenge", {
      debug: { challenge, verifier },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
