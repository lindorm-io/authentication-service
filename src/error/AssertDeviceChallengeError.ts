import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";

export class AssertDeviceChallengeError extends APIError {
  constructor(challenge: string, verifier: string) {
    super("Invalid Device Challenge", {
      debug: { challenge, verifier },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
