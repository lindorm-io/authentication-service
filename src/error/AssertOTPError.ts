import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertOTPError extends APIError {
  constructor() {
    super("Invalid OTP", {
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
