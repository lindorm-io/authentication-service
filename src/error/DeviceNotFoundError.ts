import { APIError, HttpStatus } from "@lindorm-io/core";

export class DeviceNotFoundError extends APIError {
  constructor(deviceId?: string) {
    super("Device not found", {
      debug: { deviceId },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
