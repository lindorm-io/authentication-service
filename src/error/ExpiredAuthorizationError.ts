import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";
import { Authorization } from "../entity";

export class ExpiredAuthorizationError extends APIError {
  constructor(authorization: Authorization) {
    super("Authorization is expired", {
      debug: {
        id: authorization.id,
        expires: authorization.expires,
      },
      statusCode: HttpStatus.ClientError.UNAUTHORIZED,
    });
  }
}
