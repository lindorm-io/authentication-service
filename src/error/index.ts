export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
