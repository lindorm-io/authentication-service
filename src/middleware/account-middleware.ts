import { IAuthContext } from "../typing";
import { InvalidPermissionError, LockedAccountError } from "../error";
import { Permission, Scope } from "../enum";
import { TPromise } from "@lindorm-io/core";
import { includes } from "lodash";

export const accountMiddleware = async (ctx: IAuthContext, next: TPromise<void>): Promise<void> => {
  const start = Date.now();

  const { logger, repository, token } = ctx;
  const { subject, permission, scope } = token.bearer;

  const account = await repository.account.find({ id: subject });

  if (account.permission === Permission.LOCKED) {
    throw new LockedAccountError(account.id);
  }

  if (account.permission !== permission) {
    throw new InvalidPermissionError();
  }

  const scopes = scope.split(" ");

  if (!includes(scopes, Scope.DEFAULT)) {
    throw new Error("invalid scope");
  }

  logger.debug("account found", { id: account.id });

  ctx.account = account;
  ctx.metrics = {
    ...(ctx.metrics || {}),
    account: Date.now() - start,
  };

  await next();
};
