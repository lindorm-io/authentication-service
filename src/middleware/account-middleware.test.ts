import MockDate from "mockdate";
import { InvalidPermissionError, LockedAccountError } from "../error";
import { Permission, Scope } from "@lindorm-io/jwt";
import { accountMiddleware } from "./account-middleware";
import { getTestRepository, getTestAccount, logger, resetStore } from "../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

const next = jest.fn();

describe("accountMiddleware", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      logger,
      repository: await getTestRepository(),
      token: {
        bearer: {
          subject: "be3a62d1-24a0-401c-96dd-3aff95356811",
          permission: Permission.USER,
          scope: [Scope.DEFAULT, Scope.OPENID].join(" "),
        },
      },
    };
    await ctx.repository.account.create(getTestAccount("email@lindorm.io"));
  });

  afterEach(resetStore);

  test("should successfully set account on ctx", async () => {
    await expect(accountMiddleware(ctx, next)).resolves.toBe(undefined);

    expect(ctx.account).toMatchSnapshot();
    expect(ctx.metrics.account).toStrictEqual(expect.any(Number));
  });

  test("should throw error if account is locked", async () => {
    const account = await ctx.repository.account.find({ email: "email@lindorm.io" });
    account.permission = Permission.LOCKED;
    await ctx.repository.account.update(account);

    await expect(accountMiddleware(ctx, next)).rejects.toStrictEqual(expect.any(LockedAccountError));
  });

  test("should throw error if account permission is different", async () => {
    ctx.token.bearer.permission = Permission.ADMIN;

    await expect(accountMiddleware(ctx, next)).rejects.toStrictEqual(expect.any(InvalidPermissionError));
  });
});
