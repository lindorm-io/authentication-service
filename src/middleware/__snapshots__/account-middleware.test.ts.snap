// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountMiddleware should successfully set account on ctx 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "email@lindorm.io",
  "_events": Array [],
  "_id": "accountId",
  "_otp": Object {
    "signature": null,
    "uri": null,
  },
  "_password": Object {
    "signature": null,
    "updated": null,
  },
  "_permission": "user",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
